client/src/main/scala/org/http4s/client/middleware/FollowRedirect.scala

```diff
@@ -23,23 +21,36 @@ object FollowRedirect {
                 fragment = uri.fragment orElse req.uri.fragment
               )
 
-              prepareLoop(
-                req.copy(method = method, uri = nextUri, body = EmptyBody),
-                redirects + 1
-              )
+              // We're following a redirect, so we need to dispose of the last response
+              dispose.flatMap { _ =>
+                prepareLoop(
+                  req.copy(method = method, uri = nextUri, body = EmptyBody),
+                  redirects + 1
+                )
+              }
 
             case _ => Task.now(dr)
           }
         }
 
         resp.status.code match {
-          // We cannot be sure what will happen to the request body so we don't attempt to deal with it
-          case 301 | 302 | 307 | 308 if req.body.isHalt =>
```

Comment:
I think I remember why we had the `.isHalt`: to deal with the consumption of the body. _If_ the body is a Halt, we can be confident that we won't mess things up by running it again in the redirected request.